Feature/hdr tree implementation

Jerry01020 opened the pull request:
feat: 实现 HDR-Tree 索引及相关测试集成 #63 #58

Jerry01020 committed:
初步完成HDR+Tree

Jerry01020 committed:
添加中文注释

Jerry01020 committed:
实现集成HDR Tree和简单剪枝逻辑

Jerry01020 committed:
完成PCA训练和RTree构建

Jerry01020 committed:
实现森林级剪枝

Jerry01020 committed:
基本完成逻辑功能

Jerry01020 committed:
完成HDRTree基本逻辑功能

Jerry01020 committed:
优化单元测试

Jerry01020 committed:
优化HDR_Tree效率

Jerry01020 committed:
Merge branch 'main-dev' of https://github.com/intellistream/sageFlow into main-dev

Jerry01020 committed:
优化HDR_Tree索引

Jerry01020 committed:
能通过join_datasource_modes测试

Jerry01020 committed:
HDR_Tree SIFT数据集加载测试和随机数据内存生成测试

Jerry01020 committed:
可通过现有完整流程测试的HDR_Tree

Jerry01020 committed:
add SUMMARY_OF_HDR_Tree

Jerry01020 committed:
修复HDR_Forest测试不通过的问题

ZeroJustMe commented on src/operator/join_operator.cpp:
```diff
@@ -167,7 +168,38 @@ JoinOperator::JoinOperator(std::unique_ptr<Function> &join_func,
         
-    } else if (algo == "hnsw") {
+    
+        } else if (algo == "hdrtree" || algo == "hdr_tree" || algo == "hdrforest") {
```
hdr_forest和hdr_tree是同一种方法吗？这里算子里现在用的是什么版本的方法

Jerry01020 replied:
不是同一个方法，HDR_Forest是HDR_Tree的一个集成管理模块，通过集成多个树的结果来提升查询召回率。
实际调用中使用的是HDR_Forest，只是因为复现文献使用HDR_Tree便于管理和理解

ZeroJustMe commented on src/operator/join_operator.cpp:
```diff
@@ -167,7 +168,38 @@ JoinOperator::JoinOperator(std::unique_ptr<Function> &join_func,
+        } else if (algo == "hdrtree" || algo == "hdr_tree" || algo == "hdrforest") {
+        index_kind_ = InternalIndexKind::HDR_TREE;
+        // HDRTree 推荐使用 PartitionedWindowState
+        use_shared_state_ = false; 
```
hdr_tree应该用的是共享索引吧，这里为什么用了false

Jerry01020 replied:
这里应该是当时我对这个多线程和互斥锁的内容没把握，我参考了一些别的内容：HDR_Tree 不依赖线性扫描窗口来查找邻居，索引本身是全局共享的。还有就是我记忆中原文没有考虑多线程的问题，我就先设置这个实现相对稳定的版本。
目前。我测试过低并行度下改为True对测试似乎没有影响，更多问题可能需要在服务器上测试才能观察到

ZeroJustMe commented on config/perf_join_datasource_modes.toml:
这里原先的配置先不动吧，跑hdr-tree测试的时候改成另一个配置的路径就行；还有一个ci.toml结尾的配置文件的小规模测试，可以在里面加上hdr-tree方法保证每次过ci的时候都能测到hdr-tree

Jerry01020 replied:
那我将原有的toml改回来，当前测HDR_Tree的备份，然后在ci.toml里加HDR_Tree的测试？

ZeroJustMe replied:
好

ZeroJustMe commented on src/operator/join_strategy_factory.cpp:
```diff
@@ -466,6 +466,12 @@ IndexParameters JoinStrategyFactory::getIndexParameters(const JoinStrategyConfig
             return params;
         }
+        case JoinAlgorithm::HDR_TREE: {
+            HDRForestParameters params;
```
这两个参数是啥作用，注释里为啥说是placeholder

Jerry01020 replied:
n_clusters决定Forest中树的数量，控制Forest的规模，f_sections控制森林的分段数量
应该是初始开发的时候将HDR_Tree调整为HDR_Forest时直接使用了HDR_Tree的参数配置，也有可能是系统中为了统一管理给的是HDR_Tree的id，但是上面说过了实际调用应该是Forest，补全的注释简单粗暴加了个placeholder

ZeroJustMe commented on fix_hdr_forest.py:
这个python文件是干啥的？如果是临时文件就删掉，有用的话就建一个scripts文件夹放着

Jerry01020 replied:
测试文件，下次commit删除

ZeroJustMe commented on update_hdr_forest_final.py:
这个python文件是干啥的？同另一个

Jerry01020 replied:
测试文件，下次commit删除

ZeroJustMe commented on include/index/hdr_tree.h:
```diff
@@ -34,212 +24,100 @@ class HDRTree final : public Index {
-   */
+  auto query(const VectorRecord& record, const std::vector<float>& projected, int k) -> std::vector<uint64_t>;
   auto query_for_join(const VectorRecord& record, double threshold) -> std::vector<uint64_t> override;
+  auto query_for_join(const VectorRecord& record, const std::vector<float>& projected, double threshold) -> std::vector<uint64_t>;
```
这两个query_for_join的区别是什么？实际上用的是哪个

Jerry01020 replied:
一个是Index中的基类函数的重载，函数内部进行独立的投影计算，另一个添加了已计算好的投影向量projected。
实际上使用的应该是第一个的重载，但是因为第二个是文章中描述的方法我就先保存下来了

Jerry01020 committed:
调整测试内容

ZeroJustMe commented on src/index/hdr_tree.cpp:
```diff
@@ -199,31 +227,327 @@ void HDRTree::insertToRTree(uint64_t uid, const std::vector<float>& projected) {
+std::unique_ptr<HDRTree::RTreeNode> HDRTree::insertRecursive(RTreeNode* node, uint64_t uid, const std::vector<float>& point)
+ {                                                                                                                            if (!node) {
+      std::cerr << "FATAL: insertRecursive called with null node!" << std::endl;
+      return nullptr;
```
> Error message uses `std::cerr` directly instead of the project's logging infrastructure. Should use:
> 
> ```c++
> SAGEFLOW_LOG_ERROR("HDRTree", "insertRecursive called with null node");
> ```
> 
> Same issue appears in lines 263, 296.

这个地方换成SAGEFLOW风格的LOG吧

Jerry01020 replied:
已做出调整

Jerry01020 committed:
refactor(index): optimize logging and error handling in HDR implementation

- Replace std::cerr with SAGEFLOW_LOG_ERROR in hdr_tree.cpp for better integration with project logging
- Remove unnecessary debug output in hdr_tree.cpp
- Replace iostream with logger.h in hdr_forest.cpp
- Add proper error logging for missing records and storage manager in hdr_forest.cpp
